manager/kubernetes.go

```diff
@@ -0,0 +1,155 @@
+package manager
+
+import (
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	longhorn "github.com/rancher/longhorn-manager/k8s/pkg/apis/longhorn/v1alpha1"
+	"github.com/rancher/longhorn-manager/util"
+)
+
+var (
+	pvVolumeMode = apiv1.PersistentVolumeFilesystem
+)
+
+func (m *VolumeManager) CreateAssociatedPV(name, pvName string) (v *longhorn.Volume, err error) {
+	defer func() {
+		err = errors.Wrapf(err, "unable to create associated pv for volume %v", name)
+	}()
+
+	if !util.ValidateName(name) {
+		return nil, fmt.Errorf("invalid name %v", name)
+	}
+
+	v, err = m.ds.GetVolume(name)
+	if err != nil {
+		return nil, err
+	}
+
+	// the volume had associated pv
+	if v.Status.KubernetesStatus.PVName != "" {
+		return v, fmt.Errorf("volume already had associated pv %v", v.Status.KubernetesStatus.PVName)
+	}
+
+	pv := NewPVManifest(v)
+	if pvName != "" {
+		pv.Name = pvName
+	}
+
+	_, err = m.ds.CreatePersisentVolume(pv)
+	if err != nil {
+		return nil, err
+	}
+
+	logrus.Debugf("Created associated pv for volume %v: %+v", v.Name, v.Spec)
+	return v, nil
+}
+
+func NewPVManifest(v *longhorn.Volume) *apiv1.PersistentVolume {
+	return &apiv1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: v.Name,
+		},
+		Spec: apiv1.PersistentVolumeSpec{
+			Capacity: apiv1.ResourceList{
+				apiv1.ResourceStorage: *resource.NewQuantity(v.Spec.Size, resource.BinarySI),
+			},
+			AccessModes: []apiv1.PersistentVolumeAccessMode{
+				apiv1.ReadWriteOnce,
+			},
+
+			PersistentVolumeReclaimPolicy: apiv1.PersistentVolumeReclaimDelete,
+
+			VolumeMode: &pvVolumeMode,
+
+			PersistentVolumeSource: apiv1.PersistentVolumeSource{
+				CSI: &apiv1.CSIPersistentVolumeSource{
+					Driver: "io.rancher.longhorn",
+					FSType: "ext4",
+					VolumeAttributes: map[string]string{
+						"numberOfReplicas":    string(v.Spec.NumberOfReplicas),
+						"staleReplicaTimeout": string(v.Spec.StaleReplicaTimeout),
+					},
+					VolumeHandle: v.Name,
+				},
+			},
+		},
+	}
+}
+
+func (m *VolumeManager) CreateAssociatedPVC(name, namespace, pvcName string) (v *longhorn.Volume, err error) {
+	defer func() {
+		err = errors.Wrapf(err, "unable to create associated pvc and pv for volume %v", name)
+	}()
+
+	if !util.ValidateName(name) {
+		return nil, fmt.Errorf("invalid name %v", name)
+	}
+
+	v, err = m.ds.GetVolume(name)
+	if err != nil {
+		return nil, err
+	}
+
+	ks := v.Status.KubernetesStatus
+
+	// the volume had associated pvc and pv
+	if ks.LastPVCRefAt == "" && ks.PVCName != "" {
+		return v, fmt.Errorf("volume already had associated pvc %v", ks.PVCName)
+	}
+
+	if namespace == "" {
```

Comment:
Don't set default namespace. It's not common to use the default namespace on real workload.